### PR TITLE
Ds 2149 add unit tests

### PR DESCRIPTION
**Pull request by heolin**

### CODE REVIEW

File: src/data_snack/connections/base.py [hunk 1]
```diff
@@ -1,25 +1,62 @@
...
+        :param keys: a list of keys
+        :return: a dictionary with retrieved values assigned to each key
+        """
         ...
 
     @abstractmethod
-    def set_many(self, values: Dict[Text, Union[Text, bytes]]):
+    def set_many(self, values: Dict[Text, Union[Text, bytes]]) -> List[Text]:
+        """
+        Saves multiple saves in db
```
Comment (maciejodziemczyk) [inline src/data_snack/connections/base.py]: Shouldn't be something like 
```python
Saves multiple values in db.
```
?

File: src/data_snack/connections/base.py [hunk 2]
```diff
@@ -1,25 +1,62 @@
-from abc import ABC, abstractmethod
-from typing import Text, List, Dict, Union
+from abc import abstractmethod
+from typing import Protocol, Text, List, Dict, Union, Any
 
 
-class Connection(ABC):
+class Connection(Protocol):
```
Comment (maciejodziemczyk) [inline src/data_snack/connections/base.py:5]: Is current set of methods sufficient? I saw Redis, for example, has much more. Or this is minimal needed set of methods to meet Snack requirements about database backed?

File: src/data_snack/connections/redis.py [hunk 3]
```diff
@@ -20,5 +20,5 @@ def get_many(self, keys: List[Text]) -> Dict[Text, bytes]:
     def set_many(self, values: Dict[Text, Text]):
```
Comment (maciejodziemczyk) [inline src/data_snack/connections/redis.py]: As for `memcached.py`

**Comment by patrykjarzyna** [comment]
pandas is missing as a requirement

**Comment by patrykjarzyna** [comment]
we should limit packages range

File: src/data_snack/connections/memcached.py [hunk 4]
```diff
@@ -21,5 +21,5 @@ def set_many(self, values: Dict[Text, Text]):
         failed_keys = self.connection.set_many(values)
```
Comment (maciejodziemczyk) [inline src/data_snack/connections/memcached.py:21]: 1. Why return typings are not specified? applicable to the entire file.
2. Should docstrings be added since they are in `Connection`? 
3. Why attribute typing is provided as a string? Why wouldn't you use just a memcached class that is expected here? Is this import such expensive here?
4. To be honest I am a bit confused, since in readme you mentioned that user can use any backend needed as long as it follows our Snack protocol. We have two Connection implementations which are based on (according to this PR) protocol. Shouldn't we create one ABC as an interface for concrete implementation eg. `RedisConnection`, `MemcacheConnection` and one Protocol which should be followed by particular backend e.g. Redis, Memcache? Then our ABC.Connection should be of mentioned Protocol type.

File: tests/data_snack/connections/test_memcached.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    assert result == "key"
+    connection.connection.set.assert_called_with("key", "value")
+
+
+def test_get_many(connection: MemcachedConnection) -> None:
+    """Testing getting multiple values based on a provided list of keys."""
+    connection.connection.get_many.return_value = {"key1": "value1", "key2": "value2"}
+
+    result = connection.get_many(["key1", "key2"])
+    assert result == {"key1": "value1", "key2": "value2"}
```
Comment (patrykjarzyna) [inline tests/data_snack/connections/test_memcached.py:37]: We could skip this type of asserts since it is just a mock check

Comment (patrykjarzyna) [reply tests/data_snack/connections/test_memcached.py:37]: same for other cases

File: src/data_snack/connections/base.py (hunk 2 above)
Comment (maciejodziemczyk) [inline src/data_snack/connections/base.py:5]: Why you changed it to Protocol and kept `abstractmethods` and inheritance in implementation?

Protocols are just for typing and are not used to check method implementation, also inheritance is not needed.

If we want to keep direct relation and checks we should probably stay with ABCs.

File: tests/data_snack/connections/test_memcached.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+def test_set_many(connection: MemcachedConnection) -> None:
+    """Testing setting multiple values. Values are provided in a form of a dictionary."""
+    connection.connection.set_many.return_value = []
+
+    result = connection.set_many({"key1": "value1", "key2": "value2"})
+    assert set(result) == {"key1", "key2"}
+    connection.connection.set_many.assert_called_with({"key1": "value1", "key2": "value2"})
+
+
+@pytest.mark.xfail(raises=NotImplemented)
```
Comment (patrykjarzyna) [inline tests/data_snack/connections/test_memcached.py]: Why skip this test since we could assert  
`with pytest.raises(NotImplementedError):`
and change implementation like this  
```
    def keys(self, pattern: Text) -> List[Text]:
        raise NotImplementedError
```

File: src/data_snack/serializers/dataclass.py [hunk 7]
```diff
@@ -14,14 +14,14 @@ def _serialize(entity: Entity) -> bytes:
 @dataclass
```
Comment (maciejodziemczyk) [inline src/data_snack/serializers/dataclass.py:10]: Is `_serialize` function needed since we have `_serialize` method of `DataclassSerializer`? (I can't add a comment to that line)

File: src/data_snack/serializers/dataclass.py [hunk 8]
```diff
@@ -14,14 +14,14 @@ def _serialize(entity: Entity) -> bytes:
...
 class DataclassSerializer(Serializer):
     def __post_init__(self):
-        self.entity_fields = list(self.entity_type.__dataclass_fields__.keys())
+        self.entity_fields = list(self.entity_type.__dataclass_fields__.keys())  # noqa
```
Comment (maciejodziemczyk) [inline src/data_snack/serializers/dataclass.py]: How this line works? `entity_type` is of type `Entity`, but I can't reporduce example which allows to use `__dataclass_fields__` methods. Is everything correct here in terms of typings?

EDIT:
Does it work? Entity is not a `dataclass`.

EDIT 2:
Maybe we could use `__annotations__` instead? As far as I know it doesn't see super class fields, but `Entity` is rather not supposed to use inheritance in such sense, so it won't be a problem.

EDIT 3:
I don't know if I understand correctly, `Serializer` implementation is different for different kind of `Entity` (regular class, dataclass)? If so why Dataclass serializer depends on regular class Entity instead the dataclass one? Probably that is why you need to add `# noqa` - regular class doesn't provide `__dataclass_fields__` method.

Do we want to use different kind of Entities? I mean regular class, dataclass or something else? Wouldn't it be simpler if we just assume that it should be dataclass, or regular class (which is more general so applicable to dataclass as well) - then we could remove this kind of strategy pattern from `serializer` module .

File: tests/data_snack/connections/test_memcached.py (hunk 5 above)
Comment (heolin) [reply tests/data_snack/connections/test_memcached.py:37]: Are you sure? We are mocking function called inside, so we are checking if the function is calling `self.connection.get_many` correctly

File: src/data_snack/snack.py [hunk 9]
```diff
@@ -13,66 +14,113 @@ def _build_key(type_name: Text, *keys: Text):
...
 class Snack:
+    """
+    A core interface handling saving and reading Entities from db.
+    """
+
     connection: Connection
     registry: Dict[Text, EntityRegistry] = field(default_factory=dict)
 
-    def register_entity(self, cls: EntityType, keys: List[Text], serializer: Serializer = None):
+    def register_entity(self, cls: Type[Entity], key_fields: List[Text], serializer: Serializer = None) -> None:
```
Comment (maciejodziemczyk) [inline src/data_snack/snack.py:24]: Should we rename `cls` to e.g. `entity_class` since `cls` seems to be kind a python key word (not exactly, but it is mostly used in classmethods)

File: tests/data_snack/connections/test_memcached.py (hunk 6 above)
Comment (heolin) [reply tests/data_snack/connections/test_memcached.py]: I tried that, but for some reason it causes pytest error

```
INTERNALERROR>   File "../.pyenv/versions/3.9.14/envs/data-snack/lib/python3.9/site-packages/_pytest/skipping.py", line 276, in pytest_runtest_makereport
INTERNALERROR>     if raises is not None and not isinstance(call.excinfo.value, raises):
INTERNALERROR> TypeError: isinstance() arg 2 must be a type or tuple of types
````

File: src/data_snack/snack.py [hunk 10]
```diff
@@ -13,66 +14,113 @@ def _build_key(type_name: Text, *keys: Text):
...
+        :param serializer: Serializer that is used to compress and decompress entities before saving in db
+        """
         type_name = cls.__name__
 
         if type_name in self.registry:
-            raise EntityAlreadyRegistered(f"Entity {type_name} is already registred")
+            raise EntityAlreadyRegistered(f"Entity {type_name} is already registered")
 
-        if not keys:
+        if not key_fields:
```
Comment (maciejodziemczyk) [inline src/data_snack/snack.py:37]: Is this needed? `key_fields` has no default, so the error would be raised if not passed.

File: src/data_snack/snack.py [hunk 11]
```diff
@@ -13,66 +14,113 @@ def _build_key(type_name: Text, *keys: Text):
...
 
-        for key in keys:
+        for key in key_fields:
             if key not in cls.__dataclass_fields__.keys():
                 raise WrongKeyValue(f"Key {key} not found in entity definition.")
 
         if not serializer:
             serializer = DataclassSerializer(cls)
 
         self.registry[type_name] = EntityRegistry(
```
Comment (maciejodziemczyk) [inline src/data_snack/snack.py:47]: Should we allow to remove entity or update the entity? if we want to change key_fields it is not possible now.

File: src/data_snack/connections/base.py (hunk 1 above)
Comment (heolin) [reply src/data_snack/connections/base.py]: Fixed.

File: src/data_snack/connections/base.py (hunk 2 above)
Comment (heolin) [reply src/data_snack/connections/base.py:5]: Yes, this is a minimal set of methods that are required by the Snack to access db. Also, if we will decide to support more functions, specific for certain db, then we would probably have to create a separate module/wrapper/class to do that. (like `RedisSnack`) But this would introduce an explicit dependency to `Redis` client, so we would need to do it in a sort of plugin way.

File: tests/data_snack/connections/test_memcached.py (hunk 5 above)
Comment (patrykjarzyna) [reply tests/data_snack/connections/test_memcached.py:37]: my bad, just noticed it

File: src/data_snack/connections/memcached.py (hunk 4 above)
Comment (heolin) [reply src/data_snack/connections/memcached.py:21]: 1. Yes, I will add it,
2. They are copied from the base class.
3. That would mean we need to add both of them to our requirements 
4. We don't want to use ABC here since non of those share any similarities in terms of the code base. The only think we care about is the interface of the functions, that's why we use Protocols here (in the same manner as we would use interfaces in java or c#).

File: src/data_snack/connections/base.py (hunk 2 above)
Comment (heolin) [reply src/data_snack/connections/base.py:5]: Yep, `abstractmethods` are not needed anymore

File: src/data_snack/serializers/dataclass.py (hunk 7 above)
Comment (heolin) [reply src/data_snack/serializers/dataclass.py:10]: nope, fixed

File: src/data_snack/snack.py (hunk 9 above)
Comment (heolin) [reply src/data_snack/snack.py:24]: yep

File: src/data_snack/snack.py (hunk 10 above)
Comment (heolin) [reply src/data_snack/snack.py:37]: you can pass an empty list here

File: src/data_snack/connections/memcached.py (hunk 4 above)
Comment (patrykjarzyna) [reply src/data_snack/connections/memcached.py:21]: 1. Can we add a static analysis script?

File: src/data_snack/snack.py (hunk 11 above)
Comment (heolin) [reply src/data_snack/snack.py:47]: If you want to update an entity just save the updated object with the same key values.
Removing an entity is good idea, we should cover that.

File: src/data_snack/snack.py [hunk 12]
```diff
@@ -13,66 +14,113 @@ def _build_key(type_name: Text, *keys: Text):
...
+
+        :param cls: Entity type
+        :param key_fields: a list of fields that will be used to define Entity key
+        :param serializer: Serializer that is used to compress and decompress entities before saving in db
+        """
         type_name = cls.__name__
 
         if type_name in self.registry:
-            raise EntityAlreadyRegistered(f"Entity {type_name} is already registred")
+            raise EntityAlreadyRegistered(f"Entity {type_name} is already registered")
```
Comment (patrykjarzyna) [inline src/data_snack/snack.py:35]: Should we add corner case tests for exceptions in this class?

File: src/data_snack/connections/base.py [hunk 13]
```diff
@@ -1,25 +1,58 @@
...
+    def set_many(self, values: Dict[Text, Union[Text, bytes]]) -> List[Text]:
+        """
+        Saves multiple values in db
+
+        :param values: a dictionary containing keys and corresponding values
+        :return: a list of keys succesfully saved in db
+        """
         ...
 
     @abstractmethod
```
Comment (maciejodziemczyk) [inline src/data_snack/connections/base.py]: one more `abstractmethod` here ;)

File: src/data_snack/connections/memcached.py (hunk 4 above)
Comment (maciejodziemczyk) [reply src/data_snack/connections/memcached.py:21]: 1. :ok_hand: 
2. They are copied as long as we inherit from Base (Protocol), but do we want this explicit relation? According to the [PEP 0544](https://peps.python.org/pep-0544/#explicitly-declaring-implementation) it is not a mistake, but with such design we give particular implementation access to protocol methods via `super`. I'm just curious your point of view.
3. Right
4. What do you think about creating a Protocol for `Client` and `Redis` in the files where `MemcachedConnection` or `RedisConnection` are defined? Currently the typing is a bit enigmatic since it is just a string, so the user needs to look into `Connection` implementation to see details about what attribute is expected to be. I also don't know if static analysis works with this kind of typing - I remember that there was a problem with `kedro.Pipeline` methods which return `Pipeline` (done via `-> "Pipeline"`). One would say it creates some kind of dependency since we need to update Protocol when library is updated, but we have to do it now too and the whole relation is more implicit.

File: src/data_snack/connections/memcached.py [hunk 14]
```diff
@@ -8,18 +8,18 @@
...
         return self.connection.get_many(keys)
 
-    def set_many(self, values: Dict[Text, Text]):
+    def set_many(self, values: Dict[Text, Text]) -> List[Text]:
         failed_keys = self.connection.set_many(values)
         return list(set(values.keys()) - set(failed_keys))
 
-    def keys(self, pattern: Text) -> List[bytes]:
-        raise NotImplemented()
+    def keys(self, pattern: Text) -> List[Text]:
```
Comment (maciejodziemczyk) [inline src/data_snack/connections/memcached.py:24]: Do we plan to implement this or its just a workaround since memcached doesn't provide this (I don't know)? If so, then should it be a part of `Base` protocol?

File: src/data_snack/serializers/dataclass.py [hunk 15]
```diff
@@ -7,21 +7,17 @@
...
-
-
 @dataclass
 class DataclassSerializer(Serializer):
     def __post_init__(self):
-        self.entity_fields = list(self.entity_type.__dataclass_fields__.keys())
+        self.entity_fields = list(self.entity_type.__dataclass_fields__.keys())  # noqa
 
     def _serialize(self, entity: Entity) -> bytes:
         entity_fields = entity.__dict__
```
Comment (maciejodziemczyk) [inline src/data_snack/serializers/dataclass.py:16]: shouldn't it bee `entity.__annotations__`? Doesn't `entity.__dict__` return something like (part of my toy example):
```python
mappingproxy({'__module__': '__main__',
              '__annotations__': {'a': int, 'b': float},
              '__dict__': <attribute '__dict__' of 'MyClass' objects>,
              '__weakref__': <attribute '__weakref__' of 'MyClass' objects>,
              '__doc__': 'MyClass(a: int, b: float)',
              '__init__': <function __main__.__create_fn__.<locals>.__init__(self, a: int, b: float) -> None>,
              '__repr__': <function __main__.__create_fn__.<locals>.__repr__(self)>,
              '__eq__': <function __main__.__create_fn__.<locals>.__eq__(self, other)>,
              '__hash__': None})
```
Thus, fields are in `entity.__dict__["__annotations__"]` (one level deeper)

File: src/data_snack/snack.py (hunk 11 above)
Comment (maciejodziemczyk) [reply src/data_snack/snack.py:47]: Is it possible to save updated object with the same key values? We have an if block to raise an error in that scenario :D

File: src/data_snack/snack.py [hunk 16]
```diff
@@ -13,66 +14,113 @@ def _build_key(type_name: Text, *keys: Text):
...
 
-        if not keys:
+        if not key_fields:
             raise WrongKeyValue("Entity keys cannot be empty")
 
-        for key in keys:
+        for key in key_fields:
             if key not in cls.__dataclass_fields__.keys():
```
Comment (maciejodziemczyk) [inline src/data_snack/snack.py:41]: Here, in our general `Snack`,  we use `Entity` as `dataclass`, which is a bit confusing considering how serializers are designed (dataclasses vs the regular ones, `Entity` not being a dataclass, etc, mentioned above).

File: tests/data_snack/connections/test_memcached.py (hunk 6 above)
Comment (heolin) [reply tests/data_snack/connections/test_memcached.py]: Changed to `NotImplementedError` and it works correctly

File: src/data_snack/connections/memcached.py (hunk 4 above)
Comment (heolin) [reply src/data_snack/connections/memcached.py:21]: In a separate task

File: src/data_snack/snack.py [hunk 17]
```diff
@@ -13,66 +14,113 @@ def _build_key(type_name: Text, *keys: Text):
...
                 raise WrongKeyValue(f"Key {key} not found in entity definition.")
 
         if not serializer:
             serializer = DataclassSerializer(cls)
```
Comment (maciejodziemczyk) [inline src/data_snack/snack.py:45]: I'm not sure about that.
Disadvantages:
1. Strong dependency between `Snack` and particular implementation of `Serializer`.
2. `register_entity` has low cohesion - it validates input, register it and creates a serializer.

Advantages:
1. Easy to register `Entity`, no need to pass serializer.

Since we are working with our `Entity`, maybe this possibility to pass custom serializer is not necessary, and we can stay with just `Entity`, `EntitySerializer`, `Snack`, all as dataclasses? Then it becomes one closed, simple ecosystem.

Ideas:
1. maybe we could move checks to the separated functions and just calls it here
2. if we remove custom serializers this `if not serializer` block is not needed we can just create serializer

File: src/data_snack/snack.py [hunk 18]
```diff
@@ -13,66 +14,113 @@ def _build_key(type_name: Text, *keys: Text):
...
+        :param cls: `Entity` type
+        :param key_values: a list of key values representing the entity
+        :return: a retrieved Entity object
+        """
         type_name = cls.__name__
-        _key = _build_key(type_name, *key)
+        _key = _build_key(type_name, *key_values)
         if value := self.connection.get(_key):
             return self._get_serializer(type_name).deserialize(value)
         else:
```
Comment (maciejodziemczyk) [inline src/data_snack/snack.py:101]: Is this else block needed? If block if true executes return statement so method execution is finished too.

File: src/data_snack/snack.py [hunk 19]
```diff
@@ -13,66 +14,113 @@ def _build_key(type_name: Text, *keys: Text):
...
             raise KeyError(f"Key {_key} not found.")
 
-    def get_many(self, cls: EntityType, keys: List[List[Text]]) -> List[Entity]:
+    def get_many(self, cls: Type[Entity], keys_values: List[List[Text]]) -> List[Entity]:
+        """
+        Gets list of `Entity` objects from db based on provided list of keys.
+
+        :param cls: `Entity` type
+        :param keys_values: list of keys, each list defines a set key values for one Entity object
```
Comment (maciejodziemczyk) [inline src/data_snack/snack.py:109]: is it formally a `set`? If not it is confusing.

File: src/data_snack/snack.py [hunk 20]
```diff
@@ -13,66 +14,113 @@ def _build_key(type_name: Text, *keys: Text):
...
         return self._get_serializer(type_name).deserialize(records, many=True)
 
     def set_many(self, entities: List[Entity]) -> List[Text]:
+        """
+        Saves multiple `Entity` objects in db.
+
+        :param entities: a list of Entity objects
+        :return: a list of keys generated for saved objects
+        """
         type_name = entities[0].__class__.__name__
```
Comment (maciejodziemczyk) [inline src/data_snack/snack.py:124]: We assume, `entities` contain the same subclasses of `Entity`, so if we have lets say `Car(Entity)` and `Dog(Entity)` we can't set many of them in one call - is that desired behavior? If yest then maybe we should mention it in docstring?

File: src/data_snack/wrap/entity.py [hunk 21]
```diff
@@ -0,0 +1,35 @@
+from dataclasses import dataclass, field
+from typing import Text, List, Optional, Type
+
+from data_snack.entities import Entity
+from .base import Wrap
+
+
+@dataclass
+class EntityWrap(Wrap):
+    snack: "Snack"
```
Comment (maciejodziemczyk) [inline src/data_snack/wrap/entity.py:10]: Maybe we can add protocol for `Snack` here to invert dependency? What do you think?

Overall: cool idea with this simplified interface.

File: src/data_snack/wrap/data_frame.py [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+    def set_dataframe(self, df: pd.DataFrame) -> List[Text]:
+        """
+        Uses a data frame as a source for creating multiple Entity objects.
+        Data frame needs to follow the schema defined by the Entity.
+        Each row in the data frame defines a single Entity that will be saved in the db.
+
+        :param df: a data frame with entities in a tabular form
+        :return: a list of keys created for entities
+        """
+        data = [self.entity_type(**v) for v in df.to_dict(orient="records")]
```
Comment (maciejodziemczyk) [inline src/data_snack/wrap/data_frame.py:22]: Maybe we should add a validation if all df.columns  in `entity.__annotations__`?

File: tests/data_snack/wrap/test_data_frame.py [hunk 23]
```diff
@@ -0,0 +1,77 @@
+from typing import List
```
Comment (maciejodziemczyk) [inline tests/data_snack/wrap/test_data_frame.py:1]: Should we test the methods inherited form `EntityWrap`? I know it is tested there, but we assume that this dataframe version allows to do exact the same things, but we use current implementation as a shortcut, if it changes then our tests wouldn't catch it.

File: src/data_snack/wrap/data_frame.py [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+from typing import List, Text
+
+import pandas as pd
+
+from data_snack.wrap import EntityWrap
+from data_snack.wrap.exceptions import DataFrameMissingKeyColumn
+
+
+@dataclass
+class DataFrameWrap(EntityWrap):
```
Comment (maciejodziemczyk) [inline src/data_snack/wrap/data_frame.py:11]: Do we need this class? Can't it be a feature of `EntityWrap` + you changed helper functions to this class' methods, so it works with simplified interface for wrapper only, but we have an alternative usage - non simplified. Maybe we can stay with functions and use them in `EntityWrap` as well to make this two "APIs" equal in terms of utility?

File: requirements.txt [hunk 25]
```diff
@@ -1 +1 @@
-pydantic
+pandas==1.5.1
```
Comment (maciejodziemczyk) [inline requirements.txt]: If it is a library, which is potentially used in many projects, wouldn't the life be easier if we specify a range here? I mean, currently adding snack to project requirements is more likely to generate version conflicts. Snack doesn't use too much pandas so it would be relative easy to check which versions allow to do needed operations. What do you think?

File: src/data_snack/connections/memcached.py (hunk 14 above)
Comment (heolin) [reply src/data_snack/connections/memcached.py:24]: I would want to implement it, but I don't know how. That's why I skipped it for now

File: src/data_snack/serializers/dataclass.py (hunk 15 above)
Comment (heolin) [reply src/data_snack/serializers/dataclass.py:16]: nope, `__annotations__` doesn't work with inheritance so I choose to use this approach instead
